parsing/ast_iterator.ml

```diff
@@ -638,9 +638,14 @@ let default_iterator =
       (fun this {pvb_pat; pvb_expr; pvb_attributes; pvb_loc; pvb_constraint} ->
          this.pat this pvb_pat;
          this.expr this pvb_expr;
-         Option.iter (fun ct ->
-             List.iter (iter_loc this) ct.locally_abstract_univars;
-             this.typ this ct.Parsetree.typ
+         Option.iter (function
+             | Parsetree.Simple typ -> this.typ this typ
+             | Locally_abstract_univars { vars; typ } ->
+                 List.iter (iter_loc this) vars;
+                 this.typ this typ
```

Comment:
I wonder why we distinguish these two cases `Simple` and `Locally_abstract_univars`, instead of handling Simple as an empty list of univars.

Reply:
I tend to prefer to have more code in the iterators and mappers rather than having to special case the empty list case in `Typecore` and the parsetree pretty printer.

Reply:
Naive question: does that introduce an invalid state, where `Locally_abstract_univars` has `vars = []`? Or what would this state represent?

Reply:
It does allow to introduce a node that represents ` let p : type . typ = exp` which is then equivalent to `let (p: . typ) = (exp:typ)` with extended locations used for `typ`.

That sounds like a good reason to simplify the representation to `| Constraint | Coercion `.